sympy/ntheory/residue_ntheory.py

```diff
@@ -86,33 +94,110 @@ def _primitive_root_prime_iter(p):
 
     """
     # it is assumed that p is an int
-    v = [(p - 1) // i for i in factorint(p - 1).keys()]
-    a = 2
-    while a < p:
-        for pw in v:
-            # a TypeError below may indicate that p was not an int
-            if pow(a, pw, p) == 1:
+    if p == 3:
+        yield 2
+        return
+    qs = sorted(factorint(p - 1).keys())
+    if p < 41:
+        # small case
+        if p == 23:
+            g = 5
+        elif p == 7 or p % 7 == 3:
+            # 3 is the smallest primitive root of p = 7,17,31
+            g = 3
+        else:
+            # 2 is the smallest primitive root of p = 5,11,13,19,29,37
+            g = 2
+    else:
+        v = [(p - 1) // i for i in qs]
+        for g in range(2, p):
+            if all(pow(g, pw, p) != 1 for pw in v):
```

Comment:
```suggestion
    if p < 41:
        # small case
        if p == 23:
            g = 5
        elif p == 7 or p % 7 == 3:
            # 3 is the smallest primitive root of p = 7,17,31
            g = 3
        else:
            # 2 is the smallest primitive root of p = 5,11,13,19,29,37
            g = 2
    else:
        v = [(p - 1) // i for i in (factorint(p - 1)]
        for g in range(2, p):
            if all(pow(g, pw, p) != 1 for pw in v):
```

Reply:
What if the multiplicity is more than one? Do we have to do the following, e.g. for p = 73 whose p-1 factors are {2:3, 3:2}:

```python
[(p - 1) // i**m for i,m in factorint(p - 1).items()]
```

Reply:
localizing factoring to where it is needed